### PR TITLE
ENG-1728: StorableComponent trait

### PR DESCRIPTION
**Pull request by SmartMonkey-git**

### CODE REVIEW

File: tycho-indexer/src/extractor/evm/storage.rs [hunk 1]
```diff
@@ -269,11 +269,70 @@ pub mod pg {
             todo!()
         }
 
-        fn contract_id(&self) -> ContractId {
+        fn contract_id(&self) -> storage::ContractId {
             todo!()
         }
     }
 
+    impl StorableComponent<orm::ProtocolComponent, orm::NewProtocolComponent, i64>
```
Comment (dianacarvalho1) [inline tycho-indexer/src/extractor/evm/storage.rs]: `StorableProtocolComponent` no?

File: tycho-indexer/src/extractor/evm/storage.rs [hunk 2]
```diff
@@ -269,11 +269,70 @@ pub mod pg {
...
+
+            if let Some(Value::Object(map)) = val.attributes {
+                static_attributes = map
+                    .into_iter()
+                    .map(|(key, value)| (key, Bytes::from(bytes::Bytes::from(value.to_string()))))
+                    .collect();
+            }
+
+            Ok(evm::ProtocolComponent {
+                id: evm::ContractId(val.external_id),
```
Comment (dianacarvalho1) [inline tycho-indexer/src/extractor/evm/storage.rs:347]: aah this is yet the other `ContractId` right? the one that is just a string.. this is really confusing :/

File: tycho-indexer/src/extractor/evm/storage.rs [hunk 3]
```diff
@@ -269,11 +269,70 @@ pub mod pg {
...
+            creation_ts: NaiveDateTime,
+        ) -> Result<orm::NewProtocolComponent, StorageError> {
+            let protocol_type_id = self
+                .protocol_type_id
+                .parse::<i64>()
+                .map_err(|err| {
+                    StorageError::DecodeError(
+                        "Could not parse protocol type id in StorableComponent".to_string(),
+                    )
+                })?;
```
Comment (dianacarvalho1) [inline tycho-indexer/src/extractor/evm/storage.rs:371]: why is this needed?

File: tycho-indexer/src/extractor/evm/storage.rs [hunk 4]
```diff
@@ -333,3 +392,136 @@ pub mod pg {
...
+mod test {
+    use super::*;
+    use crate::{models::ProtocolSystem, storage::postgres::orm};
+
+    use crate::{hex_bytes::Bytes, storage::StorableComponent};
+    use chrono::Utc;
+    use ethers::prelude::H160;
+    use std::str::FromStr;
+
+    #[tokio::test]
```
Comment (dianacarvalho1) [inline tycho-indexer/src/extractor/evm/storage.rs]: tokio? why?

File: tycho-indexer/src/extractor/evm/storage.rs [hunk 5]
```diff
@@ -333,3 +392,136 @@ pub mod pg {
...
+            attributes: Some(atts.clone()),
+            protocol_system_id: 0,
+            created_at: Default::default(),
+            deleted_at: None,
+            inserted_ts: Default::default(),
+            modified_ts: Default::default(),
+        };
+
+        let tokens = vec![
+            H160::from_str("0x6B175474E89094C44Da98b954EedeAC495271d0F").unwrap(),
```
Comment (dianacarvalho1) [inline tycho-indexer/src/extractor/evm/storage.rs:526]: aaaahh didn't know you could do this 👀

File: tycho-indexer/src/extractor/evm/storage.rs [hunk 6]
```diff
@@ -333,3 +392,136 @@ pub mod pg {
...
+            "{:?}",
+            std::str::from_utf8(
+                &protocol_component
+                    .static_attributes
+                    .get("key1")
+                    .unwrap()
+                    .0
+            )
+            .unwrap()
+        );
```
Comment (dianacarvalho1) [inline tycho-indexer/src/extractor/evm/storage.rs]: print?

File: tycho-indexer/src/extractor/evm/storage.rs [hunk 7]
```diff
@@ -333,3 +392,136 @@ pub mod pg {
...
+    use super::*;
+    use crate::{models::ProtocolSystem, storage::postgres::orm};
+
+    use crate::{hex_bytes::Bytes, storage::StorableComponent};
+    use chrono::Utc;
+    use ethers::prelude::H160;
+    use std::str::FromStr;
+
+    #[tokio::test]
+    async fn test_from_storage() {
```
Comment (dianacarvalho1) [inline tycho-indexer/src/extractor/evm/storage.rs]: can you give the tests better names please? say that is related to the storable protocol. there are more of these methods in this file

File: tycho-indexer/src/extractor/evm/storage.rs (hunk 3 above)
Comment (dianacarvalho1) [reply tycho-indexer/src/extractor/evm/storage.rs:371]: to convert from the enum to an i64? 🤔

Comment (SmartMonkey-git) [reply tycho-indexer/src/extractor/evm/storage.rs:371]: It is actually from String to i64.
I'm not sure why one is a string and the other an i64. Should we adjust either?

File: tycho-indexer/src/extractor/evm/storage.rs (hunk 2 above)
Comment (SmartMonkey-git) [reply tycho-indexer/src/extractor/evm/storage.rs:347]: Yeah, your are right.
There is actually a thread on this, where I suggest do merged both ContractId's

File: tycho-indexer/src/extractor/evm/storage.rs (hunk 5 above)
Comment (SmartMonkey-git) [reply tycho-indexer/src/extractor/evm/storage.rs:526]: Life can be so simple! :D

File: tycho-indexer/src/extractor/evm/storage.rs (hunk 2 above)
Comment (dianacarvalho1) [reply tycho-indexer/src/extractor/evm/storage.rs:347]: I vote yes on the merge!!

File: tycho-indexer/src/extractor/evm/storage.rs (hunk 3 above)
Comment (dianacarvalho1) [reply tycho-indexer/src/extractor/evm/storage.rs:371]: huuuuuummmmmmm I find this very confusing tbh :/ aaah wait, I think this needs to point to the row/entry ID in the database of the corresponding `ProtocolType`

File: tycho-indexer/src/extractor/evm/storage.rs (hunk 5 above)
Comment (dianacarvalho1) [reply tycho-indexer/src/extractor/evm/storage.rs:526]: well with rust I highly doubt that

File: tycho-indexer/src/extractor/evm/storage.rs (hunk 2 above)
Comment (SmartMonkey-git) [reply tycho-indexer/src/extractor/evm/storage.rs:347]: I would suggest to do it in an additional PR though, because It will involve some additional changes.

wdyt?

File: tycho-indexer/src/extractor/evm/storage.rs (hunk 3 above)
Comment (SmartMonkey-git) [reply tycho-indexer/src/extractor/evm/storage.rs:371]: Then should we change `evm::ProtocolComponent.protocol_type_id` to `evm::Protocoltype` or you think just `i64` would be sufficient?

Comment (dianacarvalho1) [reply tycho-indexer/src/extractor/evm/storage.rs:371]: no no, just the correct ID pointing to the correct object should be enough. But how would you know the ID at this point? 🤔

File: tycho-indexer/src/extractor/evm/storage.rs (hunk 2 above)
Comment (dianacarvalho1) [reply tycho-indexer/src/extractor/evm/storage.rs:347]: yes ok ok

File: tycho-indexer/src/extractor/evm/storage.rs [hunk 8]
```diff
@@ -333,3 +392,123 @@ pub mod pg {
...
+        let new_protocol_component = result.unwrap();
+
+        assert_eq!(new_protocol_component.external_id, protocol_component.id.0);
+        assert_eq!(new_protocol_component.chain_id, chain_id);
+
+        let parsed_protocol_type_id = protocol_component
+            .protocol_type_id
+            .parse::<i64>()
+            .unwrap();
+        assert_eq!(new_protocol_component.protocol_type_id, parsed_protocol_type_id);
```
Comment (tamaralipows) [inline tycho-indexer/src/extractor/evm/storage.rs]: Aren't we repeating the code here that was implemented in the method? Shouldn't we just checked a hardcoded value instead?

File: tycho-indexer/src/storage/postgres/orm.rs [hunk 9]
```diff
@@ -290,6 +290,19 @@ pub struct ProtocolComponent {
...
+#[diesel(belongs_to(ProtocolType))]
+#[diesel(belongs_to(ProtocolSystem))]
+#[diesel(table_name = protocol_component)]
+pub struct NewProtocolComponent {
+    pub external_id: String,
+    pub chain_id: i64,
+    pub protocol_type_id: i64,
+    pub protocol_system_id: i64,
+    pub attributes: Option<serde_json::Value>,
+    pub created_at: NaiveDateTime,
```
Comment (tamaralipows) [inline tycho-indexer/src/storage/postgres/orm.rs]: I don't think we add this `created_at` manually? Wouldn't this be whatever time it is inserted into the db? Or am I wrong

Comment (SmartMonkey-git) [reply tycho-indexer/src/storage/postgres/orm.rs]: No, I dont think so, but I was also unsure. What makes me doubt is that there is an `inserted_ts` and `created_at`.

Comment (tamaralipows) [reply tycho-indexer/src/storage/postgres/orm.rs]: I would remove this line from here (see how I've done it in the `NewProtocolType`)

File: tycho-indexer/src/extractor/evm/storage.rs (hunk 3 above)
Comment (SmartMonkey-git) [reply tycho-indexer/src/extractor/evm/storage.rs:371]: We would get all the ID's in the extract. Only thing we are missing to do so, is a `get_protocol_type_id` function. Which is not planned to do.
But I can do it in the future.

File: tycho-indexer/src/extractor/evm/storage.rs (hunk 2 above)
Comment (SmartMonkey-git) [reply tycho-indexer/src/extractor/evm/storage.rs:347]: https://datarevenue.atlassian.net/browse/ENG-2700

File: tycho-indexer/src/extractor/evm/storage.rs (hunk 3 above)
Comment (dianacarvalho1) [reply tycho-indexer/src/extractor/evm/storage.rs:371]: can you add a task for it please?

Comment (SmartMonkey-git) [reply tycho-indexer/src/extractor/evm/storage.rs:371]: https://datarevenue.atlassian.net/browse/ENG-2715
